### PR TITLE
DRILL-8092: Add Auto Pagination to HTTP Storage Plugin

### PR DESCRIPTION
**Pull request by jnturton**
# [DRILL-8092](https://issues.apache.org/jira/browse/DRILL-8092): Add Auto Pagination to HTTP Storage Plugin

## Description

This PR is a follow-up to #2414 and refactors the Paginator classes in that PR.  No functionality is changed.  The main goal is to base the Paginators on java.util.Iterator but some extra simplification was also possible. 
 
## Documentation
See #2414.

## Testing
See #2414.

### CODE REVIEW

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/paginator/PagePaginator.java [hunk 1]
```diff
@@ -64,16 +60,16 @@ public PagePaginator(Builder builder, int limit, int pageSize, String pageParam,
...
 
   @Override
-  public String next() {
-    if (hasLimit) {
-      return super.next();
-    } else {
-      return generateNextUrl();
-    }
+  public boolean hasNext() {
+    return !partialPageReceived && (limit < 0 || (currentPage-1) * pageSize < limit);
```
Comment (paul-rogers) [inline contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/paginator/PagePaginator.java:64]: Limit code is off. A limit of 0 should be allowed: it means schema only. Fixed in the EVF V2/Easy Format Plugin PR.

**Review by paul-rogers — COMMENTED** [review]
I missed the original PR. A natural structure is to create one Drill batch per HTTP page. This allows downstream operators to start doing their thing with the current page. If the downstream operators are in the same fragment, the processing time acts as a throttle on the HTTP server, increasing time between requests.

The one trick is that, if you're reading hugely wide data, you could, in theory, fill a vector batch before hitting the end of the current HTTP page. In this case, you'd send the first batch downstream anyway before continuing to read the next page.

**Comment by jnturton** [comment]
@paul-rogers yes, of course. I'll see if I can expand the scope of this refactoring PR to add that in.

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/paginator/PagePaginator.java [hunk 2]
```diff
@@ -47,11 +44,10 @@
    * @param pageSizeParam The API Query parameter which specifies how many results per page
    */
   public PagePaginator(Builder builder, int limit, int pageSize, String pageParam, String pageSizeParam) {
-    super(builder, paginationMode.PAGE, pageSize, limit > 0);
+    super(builder, paginationMode.PAGE, pageSize, limit);
     this.limit = limit;
     this.pageParam = pageParam;
     this.pageSizeParam = pageSizeParam;
-    this.paginatedUrls = buildPaginatedURLs();
     currentPage = 1;
```
Comment (jnturton) [inline contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/paginator/PagePaginator.java:51]: @cgivre While testing this PR today I encountered an API that starts counting pages at 0.

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/paginator/PagePaginator.java (hunk 1 above)
Comment (jnturton) [reply contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/paginator/PagePaginator.java:64]: @paul-rogers I can't see that LIMIT 0 is disallowed.  This `hasNext()` method above returns false immediately in this case and 0 records are returned for the query, is that not correct?

**Comment by jnturton** [comment]
> @paul-rogers yes, of course. I'll see if I can expand the scope of this refactoring PR to add that in.

In the paginated case, a separate batch reader is created for each page of data returned by the HTTP API.

- This plugin's JSON batch reader uses JsonLoader from EVF which will try to fill up a Drill batch for each page of API data.
- Its CSV batch reader makes direct use of the Univocity CSV parse rather than going via CompliantTextBatchReader (why?) but includes its own batch-filling loop.
- Its XML batch reader uses XmlReader from the format-xml which is also EVF-based and tries to fill up a Drill batch.

So I think the "A natural structure is to create one Drill batch per HTTP page" comment is already addressed.

**Comment by jnturton** [comment]
Since I'm not aware of any outstanding concerns I'll merge this once the CI test run passes.

**Pull request merged**